File: bindings/go/encoder.go
// Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.

package triepack

import "sort"

// Format constants (match core_internal.h).
var tpMagic = []byte{0x54, 0x52, 0x50, 0x00} // "TRP\0"

const (
	tpVersionMajor  = 1
	tpVersionMinor  = 0
	tpHeaderSize    = 32
	tpFlagHasValues = 1
)

// Control code indices.
const (
	ctrlEnd    = 0
	ctrlEndVal = 1
	ctrlSkip   = 2
	// ctrlSuffix = 3
	// ctrlEscape = 4
	ctrlBranch      = 5
	numControlCodes = 6
)

// trieEntry holds a key (as bytes) and its associated value.
type trieEntry struct {
	key []byte
	val interface{}
}

// encodeCtx holds shared context for the trie encoder.
type encodeCtx struct {
	entries   []trieEntry
	bps       int
	symbolMap [256]int
	ctrlCodes [numControlCodes]int
	hasValues bool
}

// encodeData encodes a map into the .trp binary format.
func encodeData(data map[string]interface{}) ([]byte, error) {
	// Collect entries
	entries := make([]trieEntry, 0, len(data))
	for k, v := range data {
		entries = append(entries, trieEntry{key: []byte(k), val: v})
	}

	// Sort lexicographically by key bytes
	sort.Slice(entries, func(i, j int) bool {
		a, b := entries[i].key, entries[j].key
		minLen := len(a)
		if len(b) < minLen {
			minLen = len(b)
		}
		for idx := 0; idx < minLen; idx++ {
			if a[idx] != b[idx] {
				return a[idx] < b[idx]
			}
		}
		return len(a) < len(b)
	})

	// Determine if any entries have non-null values
	hasValues := false
	for _, e := range entries {
		if e.val != nil {
			hasValues = true
			break
		}
	}

	// Symbol analysis: find unique bytes in all keys
	var used [256]bool
	for _, e := range entries {
		for _, b := range e.key {
			used[b] = true
		}
	}

	alphabetSize := 0
	for _, u := range used {
		if u {
			alphabetSize++
		}
	}
	totalSymbols := alphabetSize + numControlCodes
	bps := 1
	for (1 << uint(bps)) < totalSymbols {
		bps++
	}

	// Control codes: 0..5
	var ctrlCodes [numControlCodes]int
	for i := 0; i < numControlCodes; i++ {
		ctrlCodes[i] = i
	}

	// Symbol map: byte value -> N-bit code
	var symbolMap [256]int
	var reverseMap [256]int
	code := numControlCodes
	for i := 0; i < 256; i++ {
		if used[i] {
			symbolMap[i] = code
			if code < 256 {
				reverseMap[code] = i
			}
			code++
		}
	}

	// Build the bitstream
	w := NewBitWriter(256)

	// Write 32-byte header placeholder
	for _, b := range tpMagic {
		w.WriteU8(b)
	}
	w.WriteU8(tpVersionMajor)
	w.WriteU8(tpVersionMinor)
	flags := uint16(0)
	if hasValues {
		flags = tpFlagHasValues
	}
	w.WriteU16(flags)
	w.WriteU32(uint32(len(entries))) // num_keys
	w.WriteU32(0)                    // trie_data_offset placeholder
	w.WriteU32(0)                    // value_store_offset placeholder
	w.WriteU32(0)                    // suffix_table_offset placeholder
	w.WriteU32(0)                    // total_data_bits placeholder
	w.WriteU32(0)                    // reserved

	dataStart := w.Position() // should be 256 (32 bytes * 8)

	// Trie config: bits_per_symbol (4 bits) + symbol_count (8 bits)
	w.WriteBits(uint64(bps), 4)
	w.WriteBits(uint64(totalSymbols), 8)

	// Control code mappings
	for c := 0; c < numControlCodes; c++ {
		w.WriteBits(uint64(ctrlCodes[c]), bps)
	}

	// Symbol table
	for cd := numControlCodes; cd < totalSymbols; cd++ {
		byteVal := 0
		if cd < 256 {
			byteVal = reverseMap[cd]
		}
		writeVarUint(w, byteVal)
	}

	trieDataOffset := w.Position() - dataStart

	// Write prefix trie
	ctx := &encodeCtx{
		entries:   entries,
		bps:       bps,
		symbolMap: symbolMap,
		ctrlCodes: ctrlCodes,
		hasValues: hasValues,
	}
	valueIdx := []int{0}
	if len(entries) > 0 {
		trieWrite(ctx, w, 0, len(entries), 0, valueIdx)
	}

	valueStoreOffset := w.Position() - dataStart

	// Write value store
	if hasValues {
		for _, e := range entries {
			if err := encodeValue(w, e.val); err != nil {
				return nil, err
			}
		}
	}

	totalDataBits := w.Position() - dataStart

	w.AlignToByte()

	// Get pre-CRC buffer
	preCRCBuf := w.ToBytes()
	crcVal := computeCRC32(preCRCBuf)

	// Append CRC
	w.WriteU32(crcVal)

	outBuf := w.ToBytes()

	// Patch header offsets
	patchU32(outBuf, 12, uint32(trieDataOffset))
	patchU32(outBuf, 16, uint32(valueStoreOffset))
	patchU32(outBuf, 20, 0) // suffix_table_offset
	patchU32(outBuf, 24, uint32(totalDataBits))

	// Recompute CRC over patched data
	crcDataLen := len(outBuf) - 4
	crcVal = computeCRC32(outBuf[:crcDataLen])
	patchU32(outBuf, crcDataLen, crcVal)

	return outBuf, nil
}

// patchU32 writes a big-endian uint32 at the given byte offset.
func patchU32(buf []byte, offset int, value uint32) {
	buf[offset] = byte((value >> 24) & 0xFF)
	buf[offset+1] = byte((value >> 16) & 0xFF)
	buf[offset+2] = byte((value >> 8) & 0xFF)
	buf[offset+3] = byte(value & 0xFF)
}

// trieSubtreeSize computes the subtree size in bits (dry run).
func trieSubtreeSize(ctx *encodeCtx, start, end, prefixLen int, valueIdx []int) int {
	entries := ctx.entries
	bps := ctx.bps
	hasValues := ctx.hasValues
	bits := 0

	// Find common prefix beyond prefixLen
	common := prefixLen
	for {
		allHave := true
		ch := byte(0)
		for i := start; i < end; i++ {
			if len(entries[i].key) <= common {
				allHave = false
				break
			}
			if i == start {
				ch = entries[i].key[common]
			} else if entries[i].key[common] != ch {
				allHave = false
				break
			}
		}
		if !allHave {
			break
		}
		common++
	}

	bits += (common - prefixLen) * bps

	hasTerminal := len(entries[start].key) == common
	childrenStart := start
	if hasTerminal {
		childrenStart = start + 1
	}

	// Count children
	childCount := 0
	cs := childrenStart
	for cs < end {
		childCount++
		ch := entries[cs].key[common]
		ce := cs + 1
		for ce < end && entries[ce].key[common] == ch {
			ce++
		}
		cs = ce
	}

	if hasTerminal && childCount == 0 {
		if hasValues && entries[start].val != nil {
			bits += bps
			bits += varUintBits(valueIdx[0])
		} else {
			bits += bps
		}
		valueIdx[0]++
	} else if hasTerminal && childCount > 0 {
		if hasValues && entries[start].val != nil {
			bits += bps
			bits += varUintBits(valueIdx[0])
		} else {
			bits += bps
		}
		valueIdx[0]++
		bits += bps
		bits += varUintBits(childCount)

		cs = childrenStart
		childI := 0
		for cs < end {
			ch := entries[cs].key[common]
			ce := cs + 1
			for ce < end && entries[ce].key[common] == ch {
				ce++
			}
			if childI < childCount-1 {
				childSz := trieSubtreeSize(ctx, cs, ce, common, valueIdx)
				bits += bps
				bits += varUintBits(childSz)
				bits += childSz
			} else {
				bits += trieSubtreeSize(ctx, cs, ce, common, valueIdx)
			}
			childI++
			cs = ce
		}
	} else { // not hasTerminal, childCount > 0
		bits += bps
		bits += varUintBits(childCount)

		cs = childrenStart
		childI := 0
		for cs < end {
			ch := entries[cs].key[common]
			ce := cs + 1
			for ce < end && entries[ce].key[common] == ch {
				ce++
			}
			if childI < childCount-1 {
				childSz := trieSubtreeSize(ctx, cs, ce, common, valueIdx)
				bits += bps
				bits += varUintBits(childSz)
				bits += childSz
			} else {
				bits += trieSubtreeSize(ctx, cs, ce, common, valueIdx)
			}
			childI++
			cs = ce
		}
	}

	return bits
}

// trieWrite writes the trie subtree to the BitWriter.
func trieWrite(ctx *encodeCtx, w *BitWriter, start, end, prefixLen int, valueIdx []int) {
	entries := ctx.entries
	bps := ctx.bps
	symbolMap := ctx.symbolMap
	ctrlCodes := ctx.ctrlCodes
	hasValues := ctx.hasValues

	// Find common prefix beyond prefixLen
	common := prefixLen
	for {
		allHave := true
		ch := byte(0)
		for i := start; i < end; i++ {
			if len(entries[i].key) <= common {
				allHave = false
				break
			}
			if i == start {
				ch = entries[i].key[common]
			} else if entries[i].key[common] != ch {
				allHave = false
				break
			}
		}
		if !allHave {
			break
		}
		common++
	}

	// Write common prefix symbols
	for i := prefixLen; i < common; i++ {
		ch := entries[start].key[i]
		w.WriteBits(uint64(symbolMap[ch]), bps)
	}

	hasTerminal := len(entries[start].key) == common
	childrenStart := start
	if hasTerminal {
		childrenStart = start + 1
	}

	// Count children
	childCount := 0
	cs := childrenStart
	for cs < end {
		childCount++
		ch := entries[cs].key[common]
		ce := cs + 1
		for ce < end && entries[ce].key[common] == ch {
			ce++
		}
		cs = ce
	}

	// Write terminal
	if hasTerminal {
		if hasValues && entries[start].val != nil {
			w.WriteBits(uint64(ctrlCodes[ctrlEndVal]), bps)
			writeVarUint(w, valueIdx[0])
		} else {
			w.WriteBits(uint64(ctrlCodes[ctrlEnd]), bps)
		}
		valueIdx[0]++
	}

	if childCount == 0 {
		return
	}

	// BRANCH
	w.WriteBits(uint64(ctrlCodes[ctrlBranch]), bps)
	writeVarUint(w, childCount)

	cs = childrenStart
	childI := 0
	for cs < end {
		ch := entries[cs].key[common]
		ce := cs + 1
		for ce < end && entries[ce].key[common] == ch {
			ce++
		}

		if childI < childCount-1 {
			viCopy := []int{valueIdx[0]}
			childSz := trieSubtreeSize(ctx, cs, ce, common, viCopy)
			w.WriteBits(uint64(ctrlCodes[ctrlSkip]), bps)
			writeVarUint(w, childSz)
		}

		trieWrite(ctx, w, cs, ce, common, valueIdx)

		childI++
		cs = ce
	}
}

File: bindings/rust/src/encoder.rs
// Copyright (c) 2026 M. A. Chatterjee, BSD-2-Clause.

//! Trie encoder -- port of src/core/encoder.c.

use std::collections::HashMap;

use crate::bitstream::BitWriter;
use crate::crc32::crc32;
use crate::values::{encode_value, Value};
use crate::varint::{var_uint_bits, write_var_uint};

// Format constants (match core_internal.h)
const TP_MAGIC: [u8; 4] = [0x54, 0x52, 0x50, 0x00]; // "TRP\0"
const TP_VERSION_MAJOR: u8 = 1;
const TP_VERSION_MINOR: u8 = 0;
const TP_FLAG_HAS_VALUES: u16 = 1;

// Control code indices
const CTRL_END: usize = 0;
const CTRL_END_VAL: usize = 1;
const CTRL_SKIP: usize = 2;
const CTRL_BRANCH: usize = 5;
const NUM_CONTROL_CODES: usize = 6;

/// Internal entry: sorted key bytes + value reference index.
struct Entry {
    key: Vec<u8>,
    val_idx: usize, // index into the original values vec
}

/// Encoder context passed through recursive trie writing.
struct EncCtx {
    entries: Vec<Entry>,
    values: Vec<Value>,
    bps: usize,
    symbol_map: [u32; 256],
    ctrl_codes: [u32; NUM_CONTROL_CODES],
    has_values: bool,
}

/// Encode a HashMap<String, Value> into the .trp binary format.
///
/// Keys are sorted by their UTF-8 byte representation. Duplicate keys are
/// deduplicated (last value wins, but since HashMap has unique keys this is
/// a no-op). The output is byte-identical to the C reference encoder.
pub fn encode(data: &HashMap<String, Value>) -> Vec<u8> {
    // Collect entries as (key_bytes, value), sort by key_bytes
    let mut pairs: Vec<(Vec<u8>, &Value)> = data
        .iter()
        .map(|(k, v)| (k.as_bytes().to_vec(), v))
        .collect();
    pairs.sort_by(|a, b| a.0.cmp(&b.0));

    // Build owned values list and entry list
    let mut values: Vec<Value> = Vec::with_capacity(pairs.len());
    let mut entries: Vec<Entry> = Vec::with_capacity(pairs.len());
    for (i, (key, val)) in pairs.iter().enumerate() {
        values.push((*val).clone());
        entries.push(Entry {
            key: key.clone(),
            val_idx: i,
        });
    }

    // Determine if any entries have non-null values
    let has_values = values.iter().any(|v| !matches!(v, Value::Null));

    // Symbol analysis: find unique bytes across all keys
    let mut used = [false; 256];
    for e in &entries {
        for &b in &e.key {
            used[b as usize] = true;
        }
    }

    let alphabet_size: usize = used.iter().filter(|&&u| u).count();
    let total_symbols = alphabet_size + NUM_CONTROL_CODES;

    // Determine bits_per_symbol
    let mut bps: usize = 1;
    while (1usize << bps) < total_symbols {
        bps += 1;
    }

    // Control codes: 0..5
    let mut ctrl_codes = [0u32; NUM_CONTROL_CODES];
    for c in 0..NUM_CONTROL_CODES {
        ctrl_codes[c] = c as u32;
    }

    // Symbol map: byte value -> N-bit code
    let mut symbol_map = [0u32; 256];
    let mut reverse_map = [0u8; 256];
    let mut code = NUM_CONTROL_CODES as u32;
    for i in 0..256 {
        if used[i] {
            symbol_map[i] = code;
            if (code as usize) < 256 {
                reverse_map[code as usize] = i as u8;
            }
            code += 1;
        }
    }

    let ctx = EncCtx {
        entries,
        values,
        bps,
        symbol_map,
        ctrl_codes,
        has_values,
    };

    // Build the bitstream
    let mut w = BitWriter::new(256);

    // Write 32-byte header placeholder
    for &b in &TP_MAGIC {
        w.write_u8(b);
    }
    w.write_u8(TP_VERSION_MAJOR);
    w.write_u8(TP_VERSION_MINOR);
    let flags: u16 = if has_values { TP_FLAG_HAS_VALUES } else { 0 };
    w.write_u16(flags);
    w.write_u32(ctx.entries.len() as u32); // num_keys
    w.write_u32(0); // trie_data_offset placeholder
    w.write_u32(0); // value_store_offset placeholder
    w.write_u32(0); // suffix_table_offset placeholder
    w.write_u32(0); // total_data_bits placeholder
    w.write_u32(0); // reserved

    let data_start = w.position(); // should be 256 (32 bytes * 8)

    // Trie config: bits_per_symbol (4 bits) + symbol_count (8 bits)
    w.write_bits(bps as u64, 4);
    w.write_bits(total_symbols as u64, 8);

    // Control code mappings
    for c in 0..NUM_CONTROL_CODES {
        w.write_bits(ctrl_codes[c] as u64, bps);
    }

    // Symbol table: for each code from NUM_CONTROL_CODES..total_symbols,
    // write the byte value it maps to as a VarInt.
    for cd in NUM_CONTROL_CODES..total_symbols {
        let byte_val = if cd < 256 { reverse_map[cd] } else { 0 };
        write_var_uint(&mut w, byte_val as u64);
    }

    let trie_data_offset = (w.position() - data_start) as u32;

    // Write prefix trie
    let mut value_idx: u32 = 0;
    let num_entries = ctx.entries.len() as u32;
    if num_entries > 0 {
        trie_write(&ctx, &mut w, 0, num_entries, 0, &mut value_idx);
    }

    let value_store_offset = (w.position() - data_start) as u32;

    // Write value store
    if has_values {
        for val in &ctx.values {
            encode_value(&mut w, val);
        }
    }

    let total_data_bits = (w.position() - data_start) as u32;

    w.align_to_byte();

    // Compute CRC-32 over everything written so far
    let pre_crc_buf = w.to_bytes();
    let crc_val = crc32(&pre_crc_buf);

    // Append CRC
    w.write_u32(crc_val);

    let mut out_buf = w.to_bytes();

    // Patch header offsets (big-endian u32)
    patch_u32(&mut out_buf, 12, trie_data_offset);
    patch_u32(&mut out_buf, 16, value_store_offset);
    patch_u32(&mut out_buf, 20, 0); // suffix_table_offset
    patch_u32(&mut out_buf, 24, total_data_bits);

    // Recompute CRC over patched data (everything except last 4 bytes)
    let crc_data_len = out_buf.len() - 4;
    let crc_val = crc32(&out_buf[..crc_data_len]);
    patch_u32(&mut out_buf, crc_data_len, crc_val);

    out_buf
}

/// Patch a big-endian u32 at the given byte offset.
fn patch_u32(buf: &mut [u8], offset: usize, value: u32) {
    buf[offset] = (value >> 24) as u8;
    buf[offset + 1] = (value >> 16) as u8;
    buf[offset + 2] = (value >> 8) as u8;
    buf[offset + 3] = value as u8;
}

/// Compute the number of bits needed to encode the trie subtree for
/// entries[start..end) given that `prefix_len` characters have already been
/// consumed. This is the "dry run" pass used to compute SKIP distances.
fn trie_subtree_size(
    ctx: &EncCtx,
    start: u32,
    end: u32,
    prefix_len: usize,
    value_idx: &mut u32,
) -> u64 {
    let bps = ctx.bps;
    let mut bits: u64 = 0;

    // Find common prefix beyond prefix_len
    let common = find_common_prefix(ctx, start, end, prefix_len);

    // Emit symbols for the common prefix part
    bits += ((common - prefix_len) as u64) * (bps as u64);

    // Check if any entry terminates exactly at 'common'
    let has_terminal = ctx.entries[start as usize].key.len() == common;
    let children_start = if has_terminal { start + 1 } else { start };

    // Count children
    let child_ranges = get_child_ranges(ctx, children_start, end, common);
    let child_count = child_ranges.len();

    if has_terminal && child_count == 0 {
        // Leaf node
        if ctx.has_values && !matches!(ctx.values[ctx.entries[start as usize].val_idx], Value::Null)
        {
            bits += bps as u64; // END_VAL
            bits += var_uint_bits(*value_idx as u64);
        } else {
            bits += bps as u64; // END
        }
        *value_idx += 1;
    } else if has_terminal {
        // Terminal + children
        if ctx.has_values && !matches!(ctx.values[ctx.entries[start as usize].val_idx], Value::Null)
        {
            bits += bps as u64; // END_VAL
            bits += var_uint_bits(*value_idx as u64);
        } else {
            bits += bps as u64; // END
        }
        *value_idx += 1;

        bits += bps as u64; // BRANCH
        bits += var_uint_bits(child_count as u64);

        for (ci, &(cs, ce)) in child_ranges.iter().enumerate() {
            if ci < child_count - 1 {
                let child_sz = trie_subtree_size(ctx, cs, ce, common, value_idx);
                bits += bps as u64; // SKIP
                bits += var_uint_bits(child_sz);
                bits += child_sz;
            } else {
                bits += trie_subtree_size(ctx, cs, ce, common, value_idx);
            }
        }
    } else {
        // No terminal, just branch
        bits += bps as u64; // BRANCH
        bits += var_uint_bits(child_count as u64);

        for (ci, &(cs, ce)) in child_ranges.iter().enumerate() {
            if ci < child_count - 1 {
                let child_sz = trie_subtree_size(ctx, cs, ce, common, value_idx);
                bits += bps as u64; // SKIP
                bits += var_uint_bits(child_sz);
                bits += child_sz;
            } else {
                bits += trie_subtree_size(ctx, cs, ce, common, value_idx);
            }
        }
    }

    bits
}

/// Write the trie subtree for entries[start..end).
fn trie_write(
    ctx: &EncCtx,
    w: &mut BitWriter,
    start: u32,
    end: u32,
    prefix_len: usize,
    value_idx: &mut u32,
) {
    let bps = ctx.bps;

    // Find common prefix beyond prefix_len
    let common = find_common_prefix(ctx, start, end, prefix_len);

    // Write common prefix symbols
    for i in prefix_len..common {
        let ch = ctx.entries[start as usize].key[i];
        let code = ctx.symbol_map[ch as usize];
        w.write_bits(code as u64, bps);
    }

    // Check if any entry terminates exactly at 'common'
    let has_terminal = ctx.entries[start as usize].key.len() == common;
    let children_start = if has_terminal { start + 1 } else { start };

    // Count and locate children
    let child_ranges = get_child_ranges(ctx, children_start, end, common);
    let child_count = child_ranges.len();

    // Write terminal if present
    if has_terminal {
        if ctx.has_values && !matches!(ctx.values[ctx.entries[start as usize].val_idx], Value::Null)
        {
            w.write_bits(ctx.ctrl_codes[CTRL_END_VAL] as u64, bps);
            write_var_uint(w, *value_idx as u64);
        } else {
            w.write_bits(ctx.ctrl_codes[CTRL_END] as u64, bps);
        }
        *value_idx += 1;
    }

    if child_count == 0 {
        return;
    }

    // BRANCH
    w.write_bits(ctx.ctrl_codes[CTRL_BRANCH] as u64, bps);
    write_var_uint(w, child_count as u64);

    // For each child: optionally write SKIP, then recurse
    for (ci, &(cs, ce)) in child_ranges.iter().enumerate() {
        if ci < child_count - 1 {
            let mut vi_copy = *value_idx;
            let child_sz = trie_subtree_size(ctx, cs, ce, common, &mut vi_copy);
            w.write_bits(ctx.ctrl_codes[CTRL_SKIP] as u64, bps);
            write_var_uint(w, child_sz);
        }

        trie_write(ctx, w, cs, ce, common, value_idx);
    }
}

/// Find the common prefix length beyond `prefix_len` for entries[start..end).
fn find_common_prefix(ctx: &EncCtx, start: u32, end: u32, prefix_len: usize) -> usize {
    let mut common = prefix_len;
    loop {
        let mut all_have = true;
        let mut ch: u8 = 0;
        for i in start..end {
            let key = &ctx.entries[i as usize].key;
            if key.len() <= common {
                all_have = false;
                break;
            }
            if i == start {
                ch = key[common];
            } else if key[common] != ch {
                all_have = false;
                break;
            }
        }
        if !all_have {
            break;
        }
        common += 1;
    }
    common
}

/// Get the child ranges for entries[children_start..end) grouped by the byte
/// at position `common`. Returns a Vec of (start, end) pairs.
fn get_child_ranges(ctx: &EncCtx, children_start: u32, end: u32, common: usize) -> Vec<(u32, u32)> {
    let mut ranges = Vec::new();
    let mut cs = children_start;
    while cs < end {
        let ch = ctx.entries[cs as usize].key[common];
        let mut ce = cs + 1;
        while ce < end && ctx.entries[ce as usize].key[common] == ch {
            ce += 1;
        }
        ranges.push((cs, ce));
        cs = ce;
    }
    ranges
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encode_empty() {
        let data = HashMap::new();
        let buf = encode(&data);
        // Should be 40 bytes: 32-byte header + 4 bytes trie config + 4 bytes CRC
        // Actually the empty dict has a specific trie config section
        assert!(buf.len() >= 36);
        // Check magic
        assert_eq!(&buf[0..4], &TP_MAGIC);
        // Check version
        assert_eq!(buf[4], TP_VERSION_MAJOR);
        assert_eq!(buf[5], TP_VERSION_MINOR);
        // Check num_keys = 0
        assert_eq!(&buf[8..12], &[0, 0, 0, 0]);
    }

    #[test]
    fn test_encode_single_null() {
        let mut data = HashMap::new();
        data.insert("hello".to_string(), Value::Null);
        let buf = encode(&data);
        assert_eq!(&buf[0..4], &TP_MAGIC);
        // num_keys = 1
        assert_eq!(&buf[8..12], &[0, 0, 0, 1]);
        // flags: no values (all null)
        assert_eq!(&buf[6..8], &[0, 0]);
    }

    #[test]
    fn test_encode_single_int() {
        let mut data = HashMap::new();
        data.insert("key".to_string(), Value::UInt(42));
        let buf = encode(&data);
        assert_eq!(&buf[0..4], &TP_MAGIC);
        // num_keys = 1
        assert_eq!(&buf[8..12], &[0, 0, 0, 1]);
        // flags: has_values
        let flags = ((buf[6] as u16) << 8) | buf[7] as u16;
        assert_ne!(flags & TP_FLAG_HAS_VALUES, 0);
    }
}
